cheatsheets/File_Upload_Cheat_Sheet.md

```diff
@@ -0,0 +1,163 @@
+3. Send ZIP bombs, XML bombs (otherwise known as billion laughs attack), or simply huge files in a way to fill the server storage which hinders and damages the server's availability
+4. Overwrite an existing file on the system
+5. Client-side active content that could lead to all types of client-side attacks if the file is publicly retrievable
+6. Make the user upload a file without the user's consent, by crafting a Cross-Site Request Forgery attack (CSRF).
+
+### Public File Retrieval
+
+If the file uploaded is publicly retrievable, additional threats can be addressed:
+
+1. Public disclosure of other files
+2. Initiate a DoS attack by requesting lots of files. Requests are small, yet responses are much larger
+3. File content that could be deemed as illegal, offensive, or dangerous (_e.g._ personal data, copyrighted data, etc.) which will make you a host for such malicious files.
+
+## File Upload Protection
+
+There is no silver bullet in validating user content. Implementing a defense in depth approach is key to make the upload process harder and more locked down to the needs and requirements for the service. Implementing multiple techniques is key and recommended, as no one technique is enough to secure the service.
+
+### Extension Validation
+
+Ensure that the validation occurs after decoding the file name, and that a proper filter is set in place in order to avoid certain known bypasses, such as the following:
+
+- Double extensions, _e.g._ `.jpg.php`, where it circumvents easily the regex `\.jpg`
+- Null bytes, _e.g._ `.php%00.jpg`, where `.jpg` gets truncated and `.php` becomes the new extension
+- Generic bad regex that isn't properly tested and well reviewed. Refrain from building your own logic unless you have enough knowledge on this topic.
+
+Refer to the [Input Validation CS](Input_Validation_Cheat_Sheet.md) to properly parse and process the extension.
+
+#### Whitelist Extensions
+
+Ensure the usage of *business-critical* extensions only, without allowing any type of *non-required* extensions. If the system requires:
+
+- image upload, allow one type that is agreed upon to fit the business requirement;
+- cv upload, allow `docx` and `pdf` extensions.
+
+Based on the needs of the application, ensure the **least harmful** and the **lowest risk** file types to be used.
+
+#### Blacklist Extensions
+
+> Blacklisting extensions is a bad idea and is very dangerous. Don't do it unless you have no other choice!
+
+In order to perform this validation, specifying and identifying which patterns that could should be rejected are used in order to protect the service.
+
+### Content-Type Validation
+
+Other than defining the extension of the uploaded file, its MIME-type can be checked for a quick protection against simple file upload attacks.
+
+This can be done preferrably in a whitelist approach; otherwise, this can be done in a blacklist approach.
+
+> This should not be used on its own, as bypassing it is pretty common and easy.
```

Comment:
This should be the first sentence, and should be emphasised. Maybe something along the lines of,

`The Content-Type for uploaded files is provided by the user, and as such cannot be trusted, as it is trivial to spoof. Although it should not be relied upon for security, it provide a quick check to prevent users from unintentionally uploading files with the incorrect type.